New module: elasticache_parameter_group for modifying Elasticache cluster settings

s-hertel opened the pull request:
…ete/reset cache parameter groups.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py

##### ANSIBLE VERSION
```
ansible 2.3.0 (elasticache-parameter-group f06251c042) last updated 2017/02/03 14:33:10 (GMT -400)
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
Create, delete, modify, and reset cache parameter groups. #20842 showed interest in improved functionality of ElastiCache modules.

Sample playbook:
```
---
- hosts: localhost
  connection: local
  tasks:
    - name: "create a parameter group"
      elasticache_parameter_group:
        name: "test-group"
        group_family: "memcached1.4"
        description: "initial description"
        state: "present"
        region: "{{ region }}"
        profile: "{{ profile }}"
```

```
$ ansible-playbook my_playbooks/ec_create.yml -vv

PLAYBOOK: ec_create.yml *********************************************************************
1 plays in my_playbooks/ec_create.yml

PLAY [localhost] ****************************************************************************

TASK [Gathering Facts] **********************************************************************
ok: [localhost]
META: ran handlers

TASK [create test parameter group] **********************************************************
task path: /Users/.../ansible/my_playbooks/ec_create.yml:5
changed: [localhost] => {"changed": true, "elasticache": {"DescribeCacheParametersResponse": {"DescribeCacheParametersResult": {"CacheNodeTypeSpecificParameters": [{"AllowedValues": "1-100000", "CacheNodeTypeSpecificValues": [{"CacheNodeType": "cache.c1.xlarge", "Value": "6600"}, {"CacheNodeType": "cache.m1.large", "Value": "7100"}, {"CacheNodeType": "cache.m1.medium", "Value": "3350"}, {"CacheNodeType": "cache.m1.small", "Value": "1300"}, {"CacheNodeType": "cache.m1.xlarge", "Value": "14600"}, {"CacheNodeType": "cache.m2.2xlarge", "Value": "33800"}, {"CacheNodeType": "cache.m2.4xlarge", "Value": "68000"}, {"CacheNodeType": "cache.m2.xlarge", "Value": "16700"}, {"CacheNodeType": "cache.m3.2xlarge", "Value": "28600"}, {"CacheNodeType": "cache.m3.large", "Value": "6200"}, {"CacheNodeType": "cache.m3.medium", "Value": "2850"}, {"CacheNodeType": "cache.m3.xlarge", "Value": "13600"}, {"CacheNodeType": "cache.m4.10xlarge", "Value": "158355"}, {"CacheNodeType": "cache.m4.2xlarge", "Value": "30412"}, {"CacheNodeType": "cache.m4.4xlarge", "Value": "62234"}, {"CacheNodeType": "cache.m4.large", "Value": "6573"}, {"CacheNodeType": "cache.m4.xlarge", "Value": "14618"}, {"CacheNodeType": "cache.r3.2xlarge", "Value": "59600"}, {"CacheNodeType": "cache.r3.4xlarge", "Value": "120600"}, {"CacheNodeType": "cache.r3.8xlarge", "Value": "242600"}, {"CacheNodeType": "cache.r3.large", "Value": "13800"}, {"CacheNodeType": "cache.r3.xlarge", "Value": "29100"}, {"CacheNodeType": "cache.t1.micro", "Value": "213"}, {"CacheNodeType": "cache.t2.medium", "Value": "3301"}, {"CacheNodeType": "cache.t2.micro", "Value": "555"}, {"CacheNodeType": "cache.t2.small", "Value": "1588"}], "DataType": "integer", "Description": "The maximum configurable amount of memory to use to store items, in megabytes.", "IsModifiable": false, "MinimumEngineVersion": "1.4.5", "ParameterName": "max_cache_memory", "Source": "system"}, {"AllowedValues": "1-8", "CacheNodeTypeSpecificValues": [{"CacheNodeType": "cache.c1.xlarge", "Value": "8"}, {"CacheNodeType": "cache.m1.large", "Value": "2"}, {"CacheNodeType": "cache.m1.medium", "Value": "1"}, {"CacheNodeType": "cache.m1.small", "Value": "1"}, {"CacheNodeType": "cache.m1.xlarge", "Value": "4"}, {"CacheNodeType": "cache.m2.2xlarge", "Value": "4"}, {"CacheNodeType": "cache.m2.4xlarge", "Value": "8"}, {"CacheNodeType": "cache.m2.xlarge", "Value": "2"}, {"CacheNodeType": "cache.m3.2xlarge", "Value": "8"}, {"CacheNodeType": "cache.m3.large", "Value": "2"}, {"CacheNodeType": "cache.m3.medium", "Value": "1"}, {"CacheNodeType": "cache.m3.xlarge", "Value": "4"}, {"CacheNodeType": "cache.m4.10xlarge", "Value": "40"}, {"CacheNodeType": "cache.m4.2xlarge", "Value": "8"}, {"CacheNodeType": "cache.m4.4xlarge", "Value": "16"}, {"CacheNodeType": "cache.m4.large", "Value": "2"}, {"CacheNodeType": "cache.m4.xlarge", "Value": "4"}, {"CacheNodeType": "cache.r3.2xlarge", "Value": "8"}, {"CacheNodeType": "cache.r3.4xlarge", "Value": "16"}, {"CacheNodeType": "cache.r3.8xlarge", "Value": "32"}, {"CacheNodeType": "cache.r3.large", "Value": "2"}, {"CacheNodeType": "cache.r3.xlarge", "Value": "4"}, {"CacheNodeType": "cache.t1.micro", "Value": "1"}, {"CacheNodeType": "cache.t2.medium", "Value": "2"}, {"CacheNodeType": "cache.t2.micro", "Value": "1"}, {"CacheNodeType": "cache.t2.small", "Value": "1"}], "DataType": "integer", "Description": "The number of memcached threads to use.", "IsModifiable": false, "MinimumEngineVersion": "1.4.5", "ParameterName": "num_threads", "Source": "system"}], "Marker": null, "Parameters": [{"AllowedValues": "1-10000", "DataType": "integer", "Description": "The backlog queue limit.", "IsModifiable": false, "MinimumEngineVersion": "1.4.5", "ParameterName": "backlog_queue_limit", "ParameterValue": "1024", "Source": "system"}, {"AllowedValues": "auto,ascii", "DataType": "string", "Description": "Binding protocol.", "IsModifiable": true, "MinimumEngineVersion": "1.4.5", "ParameterName": "binding_protocol", "ParameterValue": "auto", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "If supplied, CAS operations will be disabled, and items stored will consume 8 bytes less than with CAS enabled.", "IsModifiable": true, "MinimumEngineVersion": "1.4.5", "ParameterName": "cas_disabled", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "1-1024", "DataType": "integer", "Description": "The minimum amount of space to allocate for the smallest item's key + value + flags, in bytes.", "IsModifiable": true, "MinimumEngineVersion": "1.4.5", "ParameterName": "chunk_size", "ParameterValue": "48", "Source": "system"}, {"AllowedValues": "1.01-100.00", "DataType": "float", "Description": "The growth factor controlling the size of each successive memcached chunk - each chunk will be chunk_size_growth_factor times larger than the previous chunk.", "IsModifiable": true, "MinimumEngineVersion": "1.4.5", "ParameterName": "chunk_size_growth_factor", "ParameterValue": "1.25", "Source": "system"}, {"AllowedValues": "0-64", "DataType": "integer", "Description": "Maximum amount of configuration objects that can be stored at once.", "IsModifiable": false, "MinimumEngineVersion": "1.4.14", "ParameterName": "config_max", "ParameterValue": "16", "Source": "system"}, {"AllowedValues": "1024-1048576", "DataType": "integer", "Description": "Maximum size of a configuration object.", "IsModifiable": false, "MinimumEngineVersion": "1.4.14", "ParameterName": "config_size_max", "ParameterValue": "65536", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "If enabled, flush_all command will be disabled. Applicable to 1.4.24 and higher.", "IsModifiable": true, "MinimumEngineVersion": "1.4.24", "ParameterName": "disable_flush_all", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "If supplied, when there is no more memory to store items, memcached will return an error rather than evicting items.", "IsModifiable": true, "MinimumEngineVersion": "1.4.5", "ParameterName": "error_on_memory_exhausted", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "If enabled and if lru_maintainer is enabled, will make items with an expiration time of 0 unevictable. Take caution as this will crowd out memory available for other evictable items. Will not work if lru_maintainer is not enabled. Applicable to 1.4.24 and higher.", "IsModifiable": true, "MinimumEngineVersion": "1.4.24", "ParameterName": "expirezero_does_not_evict", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "jenkins,murmur3", "DataType": "string", "Description": "Configure the hash algorithm. Currently jenkins and murmur3 are offered. Default is jenkins. Applicable to 1.4.24 and higher.", "IsModifiable": true, "MinimumEngineVersion": "1.4.24", "ParameterName": "hash_algorithm", "ParameterValue": "jenkins", "Source": "system"}, {"AllowedValues": "12-64", "DataType": "integer", "Description": "Integer multiplier for the initial size of the hash table.", "IsModifiable": false, "MinimumEngineVersion": "1.4.14", "ParameterName": "hashpower_init", "ParameterValue": "16", "Source": "system"}, {"AllowedValues": "0-86400", "DataType": "integer", "Description": "If value is set more than 0, connected clients that have idled for at least this long will be asked to close. Unit is in Seconds. Applicable to 1.4.33 and higher.", "IsModifiable": true, "MinimumEngineVersion": "1.4.33", "ParameterName": "idle_timeout", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "Try to use large memory pages.", "IsModifiable": false, "MinimumEngineVersion": "1.4.5", "ParameterName": "large_memory_pages", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "Lock down all paged memory.", "IsModifiable": false, "MinimumEngineVersion": "1.4.5", "ParameterName": "lock_down_paged_memory", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "If enabled, will enable lru_crawler at launch time. LRU crawler can be enabled at run time too. See ElastiCache documentation. Applicable to 1.4.24 and higher.", "IsModifiable": true, "MinimumEngineVersion": "1.4.24", "ParameterName": "lru_crawler", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "If enabled, will enable lru_maintainer. lru_maintainer will run a background thread which shuffles items between/within the LRU's as capacities are reached. Applicable to 1.4.24 and higher.", "IsModifiable": true, "MinimumEngineVersion": "1.4.24", "ParameterName": "lru_maintainer", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "1024-134217728", "DataType": "integer", "Description": "The size of the largest item storable in the cache, in bytes.", "IsModifiable": true, "MinimumEngineVersion": "1.4.5", "ParameterName": "max_item_size", "ParameterValue": "1048576", "Source": "system"}, {"AllowedValues": "3-65000", "DataType": "integer", "Description": "The maximum number of simultaneous connections.", "IsModifiable": false, "MinimumEngineVersion": "1.4.5", "ParameterName": "max_simultaneous_connections", "ParameterValue": "65000", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "Immediately close new connections if over maximum simultaneous connections limit.", "IsModifiable": true, "MinimumEngineVersion": "1.4.14", "ParameterName": "maxconns_fast", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "Maximize core file limit.", "IsModifiable": false, "MinimumEngineVersion": "1.4.5", "ParameterName": "maximize_core_file_limit", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "0-100000", "DataType": "integer", "Description": "The amount of memory to reserve for memcached connections and other miscellaneous overhead.", "IsModifiable": true, "MinimumEngineVersion": "1.4.5", "ParameterName": "memcached_connections_overhead", "ParameterValue": "100", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "This is an alias for enabling slab_reassign, slab_automove, lru_crawler, lru_maintainer, maxconns_fast and setting hash_algorithm=murmur3. Applicable to 1.4.33 and higher.", "IsModifiable": true, "MinimumEngineVersion": "1.4.33", "ParameterName": "modern", "ParameterValue": "1", "Source": "system"}, {"AllowedValues": "1-1000", "DataType": "integer", "Description": "The maximum number of requests per event, limits the number of requests processed for a given connection to prevent starvation.", "IsModifiable": false, "MinimumEngineVersion": "1.4.5", "ParameterName": "requests_per_event", "ParameterValue": "20", "Source": "system"}, {"AllowedValues": "0-2", "DataType": "integer", "Description": "Controls the aggressiveness of the automatic memory reassignment algorithm.", "IsModifiable": true, "MinimumEngineVersion": "1.4.14", "ParameterName": "slab_automove", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "524288,1048576", "DataType": "integer", "Description": "Lets users choose 512KB or 1MB for slab_chunk_max. Applicable to 1.4.33 and higher.", "IsModifiable": true, "MinimumEngineVersion": "1.4.33", "ParameterName": "slab_chunk_max", "ParameterValue": "524288", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "Enables the ability to reassign memory.", "IsModifiable": true, "MinimumEngineVersion": "1.4.14", "ParameterName": "slab_reassign", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "0,1", "DataType": "boolean", "Description": "If enable, users can run \"stats sizes\" command. Not recommended in production environments. Applicable to 1.4.33 and higher.", "IsModifiable": true, "MinimumEngineVersion": "1.4.33", "ParameterName": "track_sizes", "ParameterValue": "0", "Source": "system"}, {"AllowedValues": "256-1024", "DataType": "integer", "Description": "Lets users change \"watch\" command's logging buffer size. Unit is in KB. Applicable to 1.4.33 and higher.", "IsModifiable": true, "MinimumEngineVersion": "1.4.33", "ParameterName": "watcher_logbuf_size", "ParameterValue": "256", "Source": "system"}, {"AllowedValues": "64-1024", "DataType": "integer", "Description": "Lets users change \"watch\" command's worker buffer size. Unit is in KB. Applicable to 1.4.33 and higher.", "IsModifiable": true, "MinimumEngineVersion": "1.4.33", "ParameterName": "worker_logbuf_size", "ParameterValue": "64", "Source": "system"}]}, "ResponseMetadata": {"RequestId": "98fad95f-ea48-11e6-aa93-a1e683c7f8af"}}}}
META: ran handlers
META: ran handlers

PLAY RECAP **********************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0

```

s-hertel committed:
Added a new module for ElastiCache. Allows users to create/modify/delete/reset cache parameter groups.

s-hertel committed:
Commenting out something that fails right now

s-hertel committed:
Use Boto3

Add metadata

Get rid of class and just use functions

s-hertel committed:
Clean up comments

s-hertel committed:
Check that the module fails with a helpful error if required options are not provided.

s-hertel committed:
Fix facts_result

s-hertel committed:
Making pep8

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,304 @@
+        name: 'test-param-group'
+        state: 'absent'
+
+"""
```
As discussed, this needs a `RETURN` docs section.

ryansb reviewed (changes requested):
I have a concern about the use of this module -- it's not convergent on a stable state. If I (as a user) want a parameter group with certain parameters set, I think it would make more sense to take arguments like:

```
parameter_group:
  group_family: redis3.2
  name: mySpecialGroup
  state: present
  values:
    asetting: avalue
```

And then have the module decide whether updates are required, rather than having the user specify whether to update things or not.

s-hertel committed:
Allow modification to a group not yet in existence. Creates and modifies group parameter in one task.

s-hertel committed:
Fix exceptions for Boto3.

s-hertel committed:
Remove trailing whitespace.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,336 @@
+      elasticache_parameter_group:
+        name: 'test-param-group'
+        values:
+          - ['activerehashing', 'yes']
```
Instead of taking a list of `[name, value]`pairs, it'd be nice to accept a map like:

```
values:
  activerehashing: yes
  something: no
```

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,336 @@
+
+    # check that the needed requirements are available
+    if state == 'present' and exists and not values:
+        module.fail_json(msg="Group already exists. Please specify values to modify using the 'values' option or use the state "
```
For this case, the user may just want the parameter group to exist, but not care about what the params are set to (if they have some other tool to set them, for example) so instead of failing, this case should just exit as a success with the current values of the group (since you already got the info above).

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,336 @@
+        module.fail_json(msg="Group already exists. Please specify values to modify using the 'values' option or use the state "
+                         "'reset' if you want to reset all or some group parameters.")
+    elif state == 'present' and not exists and not (parameter_group_family or group_description):
+        module.fail_json(msg="Creating a group requires a group name, the group family, and a description.")
```
You don't need to check the group name here, since that's a required parameter Ansible will fail before this executes.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,336 @@
+    elif state == 'reset' and not exists:
+        module.fail_json(msg="No group %s to reset. Please create the group before using the state 'reset'." % parameter_group_name)
+    elif state == 'absent' and not exists:
+        module.fail_json(msg="No group %s to delete." % parameter_group_name)
```
This should succeed, since the goal state of `absent` has been met by the group not existing in the first place.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,336 @@
+    """ Gets the current state of the cache parameter group and creates a dict with the format: {ParameterName: [Allowed_Values, DataType, ParameterValue]}"""
+    current_info = get_info(conn, name)
+    if current_info is False:
+        return False
```
Instead of returning False, I think this function should raise an exception if the parameter group can't be found. This function only gets run to get the current parameters from an existing group, so failing silently like this will lead to problems later in the script. This value gets passed to [check_valid_modification](https://github.com/ansible/ansible/pull/21023/files#diff-2faf68f29dd9e539bec6093d1ac544e3R170) and used as a dict without a type check. It makes sense to me to have this raise an exception & fail the module if the param group can't be found.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,336 @@
+        new_value = param_value_pair[1]
+
+        # check valid modifiable parameters
+        if parameter not in modifiable_params.keys():
```
The `.key()` call here isn't needed, dictionaries respond to `in` checks by looking at their keys, and calling `keys()` will create a list (Python 2) or generator (Python 3) that will have membership checks run on it.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,336 @@
+            module.fail_json("%s is not a modifiable parameter. Valid parameters to modify are: %s." % (parameter, modifiable_params.keys()))
+
+        # check allowed datatype for modified parameters
+        str_to_type = {"integer": int, "string": str}
```
I think this will fail in Python 3, instead of str try using `six.text_type`.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,336 @@
+
+        # check allowed datatype for modified parameters
+        str_to_type = {"integer": int, "string": str}
+        if type(new_value) != str_to_type[modifiable_params[parameter][1]]:
```
Instead of type checks like this, use the `isinstance` operator like:

```
if isinstance(new_value, str_to_type[modifiable_params[parameter][1]]):`
```

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,336 @@
+                             (new_value, type(new_value), parameter, modifiable_params[parameter][1]))
+
+        # check allowed values for modifiable parameters
+        if str(new_value) not in modifiable_params[parameter][0]:
```
Again, `six.text_type` instead of `str`

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,336 @@
+
+        # check allowed values for modifiable parameters
+        if str(new_value) not in modifiable_params[parameter][0]:
+            if "-" not in modifiable_params[parameter][0]:
```
What is the meaning of `-` in this context? Special value from the API?

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,336 @@
+    try:
+        response = conn.modify_cache_parameter_group(CacheParameterGroupName=name, ParameterNameValues=format_parameters)
+    except boto.exception.ClientError as e:
+        module.fail_json(msg=e.message)
```
Be more specific in this message about what was happening when the exception happened, and remember to include the `exception=traceback.format_exc()` argument to fail_json to include the stacktrace.

ryansb commented:
Also, looks like there's an issue in the YAML docs:

```
2017-02-07 13:48:32 lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py
2017-02-07 13:48:32 ============================================================================
2017-02-07 13:48:32 TRACE:
2017-02-07 13:48:32     while parsing a block collection
2017-02-07 13:48:32       in "<string>", line 19, column 6:
2017-02-07 13:48:32              - A user-specified name for the  ... 
2017-02-07 13:48:32              ^
2017-02-07 13:48:32     expected <block end>, but found '?'
2017-02-07 13:48:32       in "elasticache_parameter_group.DOCUMENTATION", line 40, column 6:
2017-02-07 13:48:32              required: yes
2017-02-07 13:48:32              ^
2017-02-07 13:48:32 TRACE:
2017-02-07 13:48:32     while parsing a block mapping
2017-02-07 13:48:32       in "<string>", line 7, column 9:
2017-02-07 13:48:32                 "cache_parameter_group": {
2017-02-07 13:48:32                 ^
2017-02-07 13:48:32     expected <block end>, but found ','
2017-02-07 13:48:32       in "elasticache_parameter_group.RETURN", line 98, column 10:
2017-02-07 13:48:32                 },
2017-02-07 13:48:32                  ^
2017-02-07 13:48:32 ERROR: DOCUMENTATION is not valid YAML. Line 40 column 6
2017-02-07 13:48:32 ERROR: RETURN is not valid YAML. Line 98 column 10
2017-02-07 13:48:32 ERROR: Type comparison using type() found on line 175. Use isinstance() instead
```

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,336 @@
+  name:
+    description:
+     - A user-specified name for the cache parameter group.
+     required: yes
```
This needs to be un-indented by one space.

s-hertel committed:
Changed values to accept a dict instead of list.

Fixed exceptions again. All but get_info should be BotoServerErrors if they fail.

Included stacktrace in exceptions.

Used text_type instead of str for python3

s-hertel committed:
Added overlooked documentation.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,337 @@
+            "cache_parameter_group_family": "redis3.2",
+            "cache_parameter_group_name": "test-please-delete",
+            "description": "initial description"
+        },
```
This comma isn't valid YAML, since this is an indentation-delimited block.

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,337 @@
+                "content-type": "text/xml",
+                "date": "Mon, 06 Feb 2017 22:14:08 GMT",
+                "x-amzn-requestid": "947291f9-ecb9-11e6-85bd-3baa4eca2cc1"
+            },
```
Same here

ryansb reviewed (changes requested):
Looks good, but shippable is still angry.

s-hertel committed:
Fixing logic to not allow modification with illegal string values and trying to make shippable happy with YAML syntax.

s-hertel committed:
Fix YAML syntax.

s-hertel committed:
YAML syntax

ryansb reviewed (approved):
Looks great, thanks!

ryansb commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,333 @@
+
+DOCUMENTATION = """
+---
+module: elasticache
```
Oh, spoke too soon: the module name and description need to be updated since this isn't a security group module.

ryansb commented:
bot_status

s-hertel committed:
Fixing module name

ryansb commented:
bot_status

jctanner commented:
bot_status

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,333 @@
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto required for this module')
+
+    parameter_group_family  = module.params.get('group_family')
```
Don't align things like this (it violates pep8 and is annoying if a longer thing gets added later)

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,333 @@
+    module.exit_json(**facts_result)
+
+# import module snippets
+from ansible.module_utils.basic import *
```
These should be replaced along the lines of your existing  imports earlier in the file (you'll need to add `get_aws_connection_info`, `ec2_argument_spec`, `camel_dict_to_snake_dict` and any others to the explicit import list)

s-hertel replied:
Should my
```
try:
    import boto3
...
```

stay at the top of the module?

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,333 @@
+        description     = dict(required=False),
+        state           = dict(required=True),
+        values          = dict(required=False, type='dict'),
+    )
```
this shouldn't happen (two succssive brackets at same level). 

Should be

```
  argument_spec.update(
        dict(
            group_family=dict( ...
        )
    )
```

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,333 @@
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
+        group_family    = dict(required=False, type='str', choices=['memcached1.4', 'redis2.6', 'redis2.8', 'redis3.2']),
+        name            = dict(required=True, type='str'),
```
Don't align the parameters (also, as it's a dict, there should be no space between key, `=` and value)

willthames commented on lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py:
```diff
@@ -0,0 +1,333 @@
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
+        group_family    = dict(required=False, type='str', choices=['memcached1.4', 'redis2.6', 'redis2.8', 'redis3.2']),
```
`required=False` is not needed as it's the default

s-hertel replied:
Thanks much for the feedback! I will make changes.

s-hertel committed:
Making requested changes.

Remove redundant required=False since it is the default

Remove parameter alignment

Move imports

s-hertel committed:
Only importing what I need from module_utils/basic

s-hertel commented:
@willthames Should the imports remain at the bottom of the file? According the abadger's comments in https://github.com/ansible/ansible/pull/21250/files we shouldn't use this practice for new modules: "The historical note might be more of a warning: If you look at pre-existing modules, you may find imports at the bottom of the file. Do not copy that idiom into new modules as it is a historical oddity due to how modules used to be combined with libraries. Over time we're moving the imports to be in their proper place."

willthames commented:
@s-hertel yep, I did try and say that 

> These should be replaced along the lines of your existing imports earlier in the file 

but on rereading that it's not the clearest of suggestions. Basically your import statements now look correct, but in the wrong place - all imports should now be just below the end of the documentation.

s-hertel committed:
move imports to top of file below documentation

s-hertel commented:
@willthames Fixed!
